# Different Multiplication Views

### Introduction

### Vector Vector Multiplication revisited

Notice that when we perform vector vector multiplication, it continues to follow our rules of multiplying a row by a column.  For example, let's initialize our matrices $A$ and $X$, and then we can multiply some of their vectors together.

In [6]:
A = np.array([[25, 8],
              [5, 6],
              [10, 7]
             ])
X = np.array([[1, 2, 4],
              [3, 4, 5]])

Now let's select a row from A and a column from X.

In [16]:
a_1, x_1 = A[:1, :2], X[:2, :1]
a_1, x_1

(array([[25,  8]]),
 array([[1],
        [3]]))

Notice the dimensions of $a_1$ and $x_1$.

In [17]:
a_1.shape, x_1.shape

((1, 2), (2, 1))

So if multiply $a \cdot x$ here, we should get a vector of length $1x1$, or in other words a single entry.

In [18]:
a_1.dot(x_1)

array([[49]])

Now let's think about what occurs if we multiply a column by a row.

In [20]:
x_1, a_1

(array([[1],
        [3]]),
 array([[25,  8]]))

In [21]:
x_1.shape, a_1.shape

((2, 1), (1, 2))

Here, the inner dimensions match as 1 and 1, but the outer dimensions suggest that we will get a 2x2 matrix.  Let's see.

In [22]:
x_1.dot(a_1)

array([[25,  8],
       [75, 24]])

And we can see that each entry is just the corresponding row dotted with the corresponding column.

$a \cdot x^T  = \begin{pmatrix}
    1  \\
    3 \\
\end{pmatrix}  \cdot \begin{pmatrix}
25 & 8
\end{pmatrix} =  \begin{pmatrix}
25 & 8 \\ 
75 & 24 
\end{pmatrix}$
    


So we can see that our vector vector multiplication really follows the same rules of matrix matrix multiplication that we saw previously.  We also see that there really is a difference between $a^T \cdot x$ and $a \cdot x^T$.  The first results in a vector, while the second results in a matrix.

### Matrix Matrix Multiplication revisited

Now that we see that a column vector times a row vector results in a matrix, we can see a different ways of viewing matrix matrix multiplication.

$A \cdot X  =  \begin{pmatrix}
    25 & 8 \\
    5 & 6\\
    10 & 7\\
\end{pmatrix} \cdot \begin{pmatrix}
    1 & 2 & 4\\ 3 & 4 & 5 
\end{pmatrix} $

In [2]:
import numpy as np
A = np.array([[25, 8],
              [5, 6],
              [10, 7]
             ])
X = np.array([[1, 2, 4],
              [3, 4, 5]])

In [6]:
A.dot(X)

array([[ 49,  82, 140],
       [ 23,  34,  50],
       [ 31,  48,  75]])

$A \cdot X  =  \begin{pmatrix}
    25  \\
    5 \\
    10\\
\end{pmatrix} \cdot \begin{pmatrix}
    1 & 2 & 4 
\end{pmatrix}  + \begin{pmatrix}
    8  \\
    6 \\
    7\\
\end{pmatrix} \cdot \begin{pmatrix}
    3 & 4 & 5 
\end{pmatrix}$

In [27]:
A[:4, :1].dot(X[:1, :4]), A[:4, 1:2].dot(X[1:2, :4]) 

(array([[ 25,  50, 100],
        [  5,  10,  20],
        [ 10,  20,  40]]),
 array([[24, 32, 40],
        [18, 24, 30],
        [21, 28, 35]]))

And adding these two dot products up gives the resulting multiplication.

In [28]:
A[:4, :1].dot(X[:1, :4]) +  A[:4, 1:2].dot(X[1:2, :4]) 

array([[ 49,  82, 140],
       [ 23,  34,  50],
       [ 31,  48,  75]])

So another way to view matrix matrix multiplication is with each column dotted with each row, and then adding the results.

$A \cdot X  =  \begin{pmatrix}
    25 & 8 \\
    5 & 6\\
    10 & 7\\
\end{pmatrix} \cdot \begin{pmatrix}
    1 & 2 & 4\\ 3 & 4 & 5 
\end{pmatrix} = \begin{pmatrix}
    25  \\
    5 \\
    10\\
\end{pmatrix} \cdot \begin{pmatrix}
    1 & 2 & 4 
\end{pmatrix}  + \begin{pmatrix}
    8  \\
    6 \\
    7\\
\end{pmatrix} \cdot \begin{pmatrix}
    3 & 4 & 5 
\end{pmatrix} = $ 


$\begin{pmatrix}
    25 & 50 & 100 \\  5 &  10 &  20 \\ 10 &  20 & 40
\end{pmatrix}  + \begin{pmatrix}
    24 & 32 & 40 \\  18 &  24 &  30 \\ 21 &  28 & 35
\end{pmatrix} =  \begin{pmatrix}
    49 & 82 & 140 \\  23 &  34 &  50 \\ 31 &  48 & 75
\end{pmatrix}$